coreneuron/io/setup_fornetcon.cpp

```diff
@@ -0,0 +1,220 @@
+ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+THE POSSIBILITY OF SUCH DAMAGE.
+*/
+
+#include "coreneuron/coreneuron.hpp"
+#include "coreneuron/io/setup_fornetcon.hpp"
+#include "coreneuron/network/netcon.hpp"
+#include "coreneuron/nrniv/nrniv_decl.h"
+#include <map>
+
+namespace coreneuron {
+
+/**
+   If FOR_NETCON in use, setup NrnThread fornetcon related info.
+
+   i.e NrnThread._fornetcon_perm_indices, NrnThread._fornetcon_weight_perm,
+   and the relevant dparam element of each mechanism instance that uses
+   a FOR_NETCONS statement.
+
+   Makes use of nrn_fornetcon_cnt_, nrn_fornetcon_type_,
+   and nrn_fornetcon_index_ that were specified during registration of
+   mechanisms that use FOR_NETCONS.
+
+   nrn_fornetcon_cnt_ is the number of mechanisms that use FOR_NETCONS,
+   nrn_fornetcon_type_ is an int array of size nrn_fornetcon_cnt, that specifies
+   the mechanism type.
+   nrn_fornetcon_index_ is an int array of size nrn_fornetcon_cnt, that
+   specifies the index into an instance's dparam int array having the
+   fornetcon semantics.
+
+   FOR_NETCONS (args) means to loop over all NetCon connecting to this
+   target instance and args are the names of the items of each NetCon's
+   weight vector (same as the enclosing NET_RECEIVE but possible different
+   local names).
+
+   NrnThread._weights is a vector of weight groups where the number of groups
+   is the number of NetCon in this thread and each group has a size
+   equal to the number of args in the target NET_RECEIVE block. The order
+   of these groups is the NetCon Object order in HOC (the construction order).
+   So the weight vector indices for the NetCons in the FOR_NETCONS loop
+   are not adjacent.
+
+   NrnThread._fornetcon_weight_perm is an index vector into the
+   NrnThread._weight vector such that the list of indices that targets a
+   mechanism instance are adjacent.
+   NrnThread._fornetcon_perm_indices is an index vector into the
+   NrnThread._fornetcon_weight_perm to the first of the list of NetCon weights
+   that target the instance. The index of _fornetcon_perm_indices
+   containing this first in the list is stored in the mechanism instances
+   dparam at the dparam's semantic fornetcon slot. (Note that the next index
+   points to the first index of the next target instance.)
+
+**/
+
+static int* fornetcon_slot(int mtype, int instance, int fnslot, NrnThread& nt) {
+  int layout = corenrn.get_mech_data_layout()[mtype];
+  int sz = corenrn.get_prop_dparam_size()[mtype];
+  Memb_list* ml = nt._ml_list[mtype];
+  int* fn = nullptr;
+  if (layout == 1) { /* AoS */
+    fn = ml->pdata + (instance*sz + fnslot);
+  }else if (layout == 0) { /* SoA */
+    int padded_cnt = nrn_soa_padded_size(ml->nodecount, layout);
+    fn = ml->pdata + (fnslot*padded_cnt + instance);
+  }
+  return fn;
+}
+
+#if 0 // for debugging
+static void print_fornetcon_slot(const char* s, std::map<int, int> type_to_slot, NrnThread& nt) {
```

Comment:
Looks `print_fornetcon_slot` is not referenced. 

`std::map<int, int> type_to_slot` should be passed as `const std::map<int, int>& type_to_slot` to avoid copy